### ISSUE TITLE
[HOLD #26538] [$500] Settings - Using deep link, admin can request money to user workspace chat even though its not allowed

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#26523]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Open settings->workspaces->any workspace-> members->invite any member
3. Open the new chat created between workspace and user added in step 2
4. Click plus and observe that request money is not present as we are admin
5. In the URL, replace '/r/' with '/request/new/' and observe that it allows to open request money page
6. Complete the process and observe that app raises a request

## Expected Result:
App should display 'Hmm its not here' page for request money by deep link by admin in workspace chat as request money option is not available for admin

## Actual Result:
App allows to request money by deep link by admin in workspace chat even though request money option is not available for admin

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.61-3

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/bd1189fd-151c-429a-b57b-d976626d65d9

https://github.com/Expensify/App/assets/115492554/c19de3ee-768d-4c47-be61-c2971efc939a

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Dhanashree-Sawant

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692943365255729

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fea0f36079525f3d</li>
        <li>Upwork Job ID: 1698569709503803392</li>
        <li>Last Price Increase: 2023-10-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sophiepintoraetz</details>

**Comment by esh-g** [comment]
# Proposal 

## Please re-state the issue
The request money deep link has no checks to ensure that the action is allowed or not.

## What is the root cause of this issue?
The request money page `MoneyRequestSelectorPage` has no logic to check if the requesting action is allowed for the particular report.

## What changes should be made in order to fix this?
We already have a method called `ReportUtils.getMoneyRequestOptions()` which returns the valid money request options for a particular report that we are using in the `+` button in the composer to display valid options.
https://github.com/Expensify/App/blob/c16995257c738761206ea07645e863ee717ecbc0/src/libs/ReportUtils.js#L2977-L2996

We should use this function in `MoneyRequestSelectorPage` as well. Since, we are already using `FullScreenNotFoundView` in the page already:
https://github.com/Expensify/App/blob/c16995257c738761206ea07645e863ee717ecbc0/src/pages/iou/MoneyRequestSelectorPage.js#L90
1. We need to check if the passed `iouType` is allowed for the current `report`. Therefore, we should modify the condition to the following:
```js
const isInvalid = !IOUUtils.isValidMoneyRequestType(iouType) || (props.report && !_.includes(ReportUtils.getMoneyRequestOptions(props.report, reportParticipantIDs, props.betas), iouType));
...
<FullPageNotFoundView shouldShow={isInvalid}> 
```

2. We also need to pass the `reportParticipantIDs` so we can add that to the body of the component as a memo:
```js
const reportParticipantIDs = useMemo(
    () => _.without(lodashGet(props.report, 'participantAccountIDs', []), props.currentUserPersonalDetails.accountID),
    [props.currentUserPersonalDetails.accountID, props.report],
);
```

3. To get the relevant parameters for `getMoneyRequestOptions()`, we can modify the `HOCs` and add some properties to the `withOnyx` `HOC` as well as add the `withCurrentUserPersonalDetails` `HOC`, like this:
```js
export default compose(
    withCurrentUserPersonalDetails,
    withOnyx({
        selectedTab: {
            key: `${ONYXKEYS.SELECTED_TAB}_${CONST.TAB.RECEIPT_TAB_ID}`,
        },
        report: {
            key: ({route}) => `${ONYXKEYS.COLLECTION.REPORT}${route.params.reportID}`,
        },
        betas: {
            key: ONYXKEYS.BETAS,
        }
    })
)(MoneyRequestSelectorPage);
```

**Comment by sophiepintoraetz** [comment]
![2023-09-04_17-31-33 (1)](https://github.com/Expensify/App/assets/42940078/9038b920-9e08-47f2-995e-379a158a6846)

**Comment by esh-g** [comment]
@ntdiary I think a combination of backend and frontend are required. This is how we handled the issue of accessing things like workspace settings for non-admins. Like in this issue: https://github.com/Expensify/App/issues/22013

**Comment by sophiepintoraetz** [comment]
>if there are other normal flows that would cause a user to hit this issue? 

I appreciate it's not a normal flow but given the world of finance, these are the sort of back doors you want to shore up, so let's fix it. If it is on the back end as well, I'll assign the `internal` label.

**Comment by sophiepintoraetz** [comment]
Hey @pecanoro - similar to this issue [here,](https://github.com/Expensify/App/issues/22013) we believe that this requires a fix on both the back end and front end - so not sure if you want to take care of both or just the back end?

**Comment by pecanoro** [comment]
Can someone summarize what I need to block on the back-end? Blocking admins from making money requests on workspace chats?

**Comment by esh-g** [comment]
@ntdiary In that case, we can subscribe to the `ONYXKEYS.SESSION` in the `withOnyx` HOC and access `accountID` like this: `props.session.accountID`

Infact, this is how we are getting the accountID in the `currentUserPersonalDetails` HOC

**Comment by pecanoro** [comment]
Asked internally because I am not sure what's the current behavior

**Comment by pecanoro** [comment]
Ok, figured this out, I am going to move to weekly so I can take care of this bug during the week but it's not extremely urgent. Once we have the back-end fix, we can implement the front-end ones.

**Comment by sophiepintoraetz** [comment]
Actually moving to weekly

**Comment by esh-g** [comment]
> Ok, figured this out, I am going to move to weekly so I can take care of this bug during the week but it's not extremely urgent. Once we have the back-end fix, we can implement the front-end ones.

@pecanoro why not do them in parallel? Because I don't this the front end PR will interfere with the backend one in any way...

**Comment by sophiepintoraetz** [comment]
@pecanoro - what's happening here?

**Comment by esh-g** [comment]
@ntdiary Would like to hear your thoughts as to if we can proceed with frontend and backend simultaneously

**Comment by pecanoro** [comment]
Basically if we proceed with both there is a high chance that I will forget to fix it in the back-end since I currently have a high amount of workload. But sure, let's try both at the same time, though I don't think I can handle this until next week.

**Comment by pecanoro** [comment]
@ntdiary Was there any approved proposal?

**Comment by sophiepintoraetz** [comment]
Oh, sounds like we're holding this issue then?

**Comment by pecanoro** [comment]
@ntdiary I am not sure if that thread ended up agreeing on a path forward. Am I missing something?

**Comment by sophiepintoraetz** [comment]
Where did we end up here @ntdiary?

**Comment by sophiepintoraetz** [comment]
(Sorry, Im out in the depths of Indonesia and wifi is verrrry slow)

**Comment by sophiepintoraetz** [comment]
If they ask why, just say that there have not been sufficient funds over the last three months, and they have had negative days in balance. We cannot give more specifics than that, as we don't want to give away our risk protection mechanism!

**Comment by techievivek** [comment]
Coming from here: https://github.com/Expensify/App/issues/29367#issuecomment-1759919616 we also want to prevent send money in group chat or other flows except global create and DM 1:1.

**Comment by pecanoro** [comment]
@ntdiary I think Tim closed the discussion here with a good proposal to move forward, do you think we could apply to this problem as well?

**Comment by pecanoro** [comment]
Ahh got it, I will put in on hold as well/

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by pecanoro** [comment]
I see that requesting money seems to be available now for admins both ways (using the + and the link), so should it actually be allowed? I am confused about what's the desired behavior. @sophiepintoraetz Can you maybe bring it up in Slack? It seems admins can request money just fine like anyone else.

**Comment by sophiepintoraetz** [comment]
Yes, I am going to close this issue out - Admins will be able to request money as a part of the invoicing/bill pay flow we are building.
